Enable smart-tab command in web extension

hohy opened the pull request:
## Changes
 - Smart Tab command enabled in web extension
 - Removed `fs` module from integration tests (replaced with VSCode API)
 - Telemetry key is injected directly into the code by CI build pipeline, so it's not needed to load it from filesystem from the `resources/TELEMETRY_KEY` file. The `TELEMETRY_INSTRUMENTATION_KEY` secret is used as source of the key. It should be already defined in the repository for the `master` branch as it was also used in previous version, but someone with admin access to the repository should check it.
 - Smart tab settings was separated from main Settings module to it's own module, so ESBuild can easily tree shake all settings code that uses node `fs` module and is not used by the web extension.
 - Made `TC288736` test more stable - it started to fail for no reason on my new machine.
 - Added new test for the `Settings` module to pass coverage threshold

## How Has This Been Tested?
Added new web extension integration test for the smart tab feature
- [x] `clients/cobol-lsp-vscode-extension/src/web/test/suite/smart-tab.test.ts`

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

hohy committed:
Make smart-tab command working in web extension

removed fs module from integration tests

await configuration change

Inject telemetry key from CI secret instead of reading it from fs

Separate smart tab setting into separate module for easier tree shaking

Fix test

Make 80 char limit test more stable

cleanup

correct working directory for inject step

ishche commented:
I would expect some delays with putting TELEMETRY_KEY into repo secrets, maybe it make sense to make separate PRs so other changes are not blocked by it.

hohy commented:
> I would expect some delays with putting TELEMETRY_KEY into repo secrets, maybe it make sense to make separate PRs so other changes are not blocked by it.

~~Hmm, good to know 👍 I guess we can put the key into the `constants.ts` file for now and remove it later once the secret is ready.~~

After more digging it seems there is already `TELEMETRY_INSTRUMENTATION_KEY` secret configured in the repo. So I'll try to use it.

hohy committed:
Use `TELEMETRY_INSTRUMENTATION_KEY` secret

Move inject steps after unit test run

Add test to pass coverage threshold

Move injection step to packageNative job

Make key injection work on windows as well

Use powershell injection for all platforms

hohy commented:
@slavek-kucera @VitGottwald Ready for review, thanks!

slavek-kucera commented on .github/workflows/release.yml:
```diff
@@ -46,5 +46,4 @@ jobs:
       env:
-        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-        THE_KEY: ${{ secrets.TELEMETRY_INSTRUMENTATION_KEY }}
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
I think that at this point the `release` job produces a package without a telemetry key.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/services/SmartTabSettings.ts:
```diff
@@ -0,0 +1,53 @@
+  const config = vscode.workspace.getConfiguration().get(SETTINGS_TAB_CONFIG);
+  let settings = new TabSettings([], new TabRule([0, 6, 7, 11], 72));
+  if (Array.isArray(config)) {
+    const stops = config as number[];
```
add type checks in a follow-up PR?

hohy replied:
I added types and input validation for the configuration 👍

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/test/suite/testHelper.ts:
```diff
@@ -79,6 +79,13 @@ export async function showDocument(workspace_file: string) {
 }
 
+export async function openUntitledDocument(languageId = LANGUAGE_ID) {
+  await vscode.commands.executeCommand(
+    "workbench.action.files.newUntitledFile",
+    { languageId },
+  );
```
I think it would be better to `openTextDocument` and `showTextDocument` and return the editor directly.

hohy replied:
It's good idea to return the editor directly, but why is `openTextDocument` and `showTextDocument` better than `newUntitledFile`?

slavek-kucera replied:
Because showTextDocument returns the editor.

hohy replied:
Done 👍

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/test/suite/testHelper.ts:
```diff
@@ -198,27 +205,15 @@ export function updateConfig(configFileName: string) {
     "settings",
     configFileName,
   );
```
Uri could be used directly.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/test/suite/testHelper.ts:
```diff
@@ -198,27 +205,15 @@ export function updateConfig(configFileName: string) {
-      fs.copyFileSync(origin, dest);
-    }
-  }
+export async function deleteFile(path: string) {
```
is it even used?

hohy replied:
It's not, removed 👍

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts:
```diff
@@ -60,12 +60,27 @@ suite("Integration Test Suite", function () {
+    await helper.waitFor(
+      () => vscode.languages.getDiagnostics(editor.document.uri).length > 0,
+    );
+    const diagnosticsCount = vscode.languages.getDiagnostics(
```
is there a variability in how many diagnostics we get?

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts:
```diff
@@ -295,9 +310,10 @@ suite("Integration Test Suite", function () {
-      helper.recursiveCopySync(
-        path.join(getWorkspacePath(), user1FilePath),
-        path.join(getWorkspacePath(), extSrcUser1FilePath),
+      await vscode.workspace.fs.copy(
+        vscode.Uri.file(path.join(getWorkspacePath(), user1FilePath)),
+        vscode.Uri.file(path.join(getWorkspacePath(), extSrcUser1FilePath)),
+        { overwrite: true },
```
use Uri directly with joinPath?

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/web/test/suite/smart-tab.test.ts:
```diff
@@ -0,0 +1,116 @@
+    });
+
+    test("Simulates tab presses", async function () {
+      await helper.openUntitledDocument();
+
+      const editor = vscode.window.activeTextEditor!;
```
```suggestion
      const editor = await helper.openUntitledDocument();;
```

slavek-kucera commented on release/release-it-release.json:
```diff
@@ -6,9 +6,7 @@
-      "echo $THE_KEY > resources/TELEMETRY_KEY",
-      "npx vsce package",
-      "git checkout -- resources/TELEMETRY_KEY"
+      "npx vsce package"
```
nobody provides the telemetry key now.

slavek-kucera commented on Jenkinsfile:
```diff
@@ -188,24 +188,7 @@ pipeline {
                     }
                 }
-
-                stage('Update telemetry key') {
```
Is the `Jenkinsfile` still used?

hohy replied:
No, it's not used, removed 👍

slavek-kucera commented on .github/workflows/build.yml:
```diff
@@ -374,6 +370,19 @@ jobs:
+      - name: Inject telemetry key to COBOL LS extension
+        shell: pwsh
+        run: |
+          $telKey = node -e "console.log(Buffer.from('${{ secrets.TELEMETRY_INSTRUMENTATION_KEY }}').toString('base64'))"
```
a simple `sed -i` command?

hohy replied:
That was my first choice, but this step runs on all different platforms and the `sed` version mac is not compatible with linux `sed`. This powershell version seems to be running on all platforms reliably.

hohy committed:
Use vscode uri instead of path

hohy committed:
Revert to static version of test

hohy committed:
Removed unused function

hohy committed:
Improved typing and configuration parsing

hohy committed:
Return editor when opening untitled file

hohy committed:
Use sed to inject telemetry key

hohy committed:
Added test to pass coverage threshold

hohy committed:
fix sed command

hohy committed:
Make sed work on all platforms

hohy committed:
Removed unused Jenkinsfile

hohy committed:
Inject telemetry key in release workflow

slavek-kucera commented on release/release-it-release.json:
```diff
@@ -6,9 +6,9 @@
     "after:bump": [
       "git add ../../CHANGELOG.md",
-      "echo $THE_KEY > resources/TELEMETRY_KEY",
+      "sed -i 's/const TELEMETRY_KEY_ENCODED: string = TELEMETRY_DEFAULT_CONTENT/const TELEMETRY_KEY_ENCODED: string = '$THE_KEY'/' src/services/reporter/TelemetryReporterImpl.ts",
```
```suggestion
      "sed -i \"s/const TELEMETRY_KEY_ENCODED: string = TELEMETRY_DEFAULT_CONTENT/const TELEMETRY_KEY_ENCODED: string = '$THE_KEY'/\" src/services/reporter/TelemetryReporterImpl.ts",
```

hohy committed:
Fixed escaping

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

hohy committed:
Use openTextDocument instead of newUntitledFile command

hohy committed:
Empty string is considered invalid telemetry key

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/test/suite/testHelper.ts:
```diff
@@ -186,39 +196,21 @@ export function range(p0: vscode.Position, p1: vscode.Position): vscode.Range {
-  const settinsFileLoc = path.join(
-    vscode.workspace.workspaceFolders![0].uri.fsPath,
+  const settingsFileLoc = vscode.Uri.joinPath(
+    vscode.Uri.file(vscode.workspace.workspaceFolders![0].uri.fsPath),
```
```suggestion
    vscode.workspace.workspaceFolders![0].uri,
```

slavek-kucera commented on release/release-it-release.json:
```diff
@@ -6,9 +6,9 @@
     "after:bump": [
       "git add ../../CHANGELOG.md",
-      "echo $THE_KEY > resources/TELEMETRY_KEY",
+      "sed -i \"s/const TELEMETRY_KEY_ENCODED: string = TELEMETRY_DEFAULT_CONTENT/const TELEMETRY_KEY_ENCODED: string = '$THE_KEY'/\" src/services/reporter/TelemetryReporterImpl.ts",
```
will this again break on macos?

hohy replied:
Release job runs only on `ubuntu-latest`.

hohy replied:
I think the release job is actually not used at all any more. It should create commit in commit history to mark the release and the last release commit I found is for release 1.1.2 from 2 years ago.

hohy committed:
use uri direclly

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

ishche commented:
So how this PR will affect release process?

hohy commented:
> So how this PR will affect release process?

@ishche It should not affect it at all. The telemetry key is injected from the secret configured in the repo, same as before. It's just not injected into the `resources/TELEMETRY_KEY` file but directly into the extension code.

ishche commented:
> > So how this PR will affect release process?
> 
> @ishche It should not affect it at all. The telemetry key is injected from the secret configured in the repo, same as before. It's just not injected into the `resources/TELEMETRY_KEY` file but directly into the extension code.

Thanks for clarification :)
